### PR TITLE
feat(inputs.aliyunrds): Add plugin

### PR DESCRIPTION
**Pull request by ZPascal**
# Add Alibaba Cloud RDS Plugin for Telegraf

This PR is a follow-up to [#17745](https://github.com/influxdata/telegraf/pull/17745) and introduces support for monitoring Alibaba Cloud RDS (Relational Database Service) instances in Telegraf.

## Summary

The new `aliyunrds` input plugin enables the collection of performance metrics from Alibaba Cloud RDS instances across multiple regions with automatic instance discovery capabilities.

### Core Functionality
- **Multi-Region Support**: Monitor RDS instances across all 21 Alibaba Cloud regions
- **Automatic Discovery**: Periodically discovers RDS instances in configured regions (default: every 1 minute)
- **Multiple Database Engines**: Supports MySQL, PostgreSQL, SQL Server, MariaDB, and PPAS
- **Flexible Configuration**: Monitor all discovered instances or specify specific instance IDs

### Metrics Collection
- Collects performance metrics from RDS Performance API
- Supports common metrics such as:
  - Sessions (MySQL_Sessions)
  - QPS/TPS (MySQL_QPS, MySQL_TPS)
  - IOPS (MySQL_IOPS)
  - Network traffic (MySQL_NetworkInNew, MySQL_NetworkOutNew)
  - Resource usage (CpuUsage, MemoryUsage, DiskUsage, ConnectionUsage)

### Authentication
Supports multiple authentication methods (evaluated in order):
1. RAM RoleArn credential
2. AccessKey STS token credential
3. AccessKey credential
4. ECS RAM Role credential
5. RSA keypair credential
6. Environment variables credential
7. Instance metadata credential

### Rate Limiting & Performance
- Configurable rate limiting (default: 200 requests/second)
- Discovery uses 20% of the configured rate limit
- Adjustable discovery interval and metric collection period
- Concurrent metric collection with proper synchronization

### Files Added
- `plugins/inputs/aliyunrds/aliyunrds.go` - Main plugin implementation
- `plugins/inputs/aliyunrds/discovery.go` - Instance discovery tool
- `plugins/inputs/aliyunrds/aliyunrds_test.go` - Comprehensive test suite
- `plugins/inputs/aliyunrds/sample.conf` - Sample configuration
- `plugins/inputs/aliyunrds/README.md` - Complete documentation
- `plugins/inputs/all/aliyunrds.go` - Plugin registration

### Output Format
Metrics are formatted as `{measurement},{tags} {fields} {timestamp}`:
- **Measurement**: `aliyunrds`
- **Tags**: `instanceId`, `region`, plus discovery tags (RegionId, DBInstanceType, Engine, EngineVersion, DBInstanceDescription)
- **Fields**: Named as `{metric_name}_{statistic}` in snake_case (e.g., `my_sql_sessions_sessions`, `cpu_usage_value`)

### Configuration Options
- `regions`: List of regions to monitor (defaults to all 21 regions)
- `period`: Aggregation period (minimum 1 minute)
- `delay`: Collection delay to account for metric availability
- `discovery_interval`: How often to run discovery (default: 1 minute)
- `ratelimit`: API request rate limit (default: 200/second)
- `metrics.names`: List of metric names to collect
- `metrics.instances`: Optional list of specific instance IDs to monitor

## Testing
Includes comprehensive test coverage:
- Unit tests for all core functions
- Mock RDS client for testing
- Tests for error handling, edge cases, and data parsing
- Discovery tool tests
- Multi-region and multi-instance scenarios

## Documentation
- Complete README with usage examples
- Sample configuration file with detailed comments
- Metric reference with links to official Alibaba Cloud documentation

## Checklist

- [x] Updated README.md with new configuration option
- [x] Updated sample.conf with new configuration option  
- [x] Added initial unit tests for new functionality
- [x] Complete unit test coverage (pending approval of approach)
- [x] Code follows project style guidelines
- [x] Backward compatible with existing configurations
- [x] End-to-end test of the plugins
- [x] AI-generated code used in this PR follows the [InfluxData Policy on AI-Generated Code Contributions](https://www.influxdata.com/ai-generated-code-contributions-policy)

### CODE REVIEW

**Comment by ZPascal** [comment]
Hi @srebhan, can you please check the pull request?

File: plugins/inputs/aliyuncms/aliyuncms.go [hunk 1]
```diff
@@ -12,14 +12,14 @@ import (
...
 	"github.com/aliyun/alibaba-cloud-sdk-go/sdk"
-	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/auth/credentials/providers"
 	"github.com/aliyun/alibaba-cloud-sdk-go/services/cms"
 	"github.com/jmespath/go-jmespath"
 
 	"github.com/influxdata/telegraf"
 	"github.com/influxdata/telegraf/config"
 	"github.com/influxdata/telegraf/internal"
 	"github.com/influxdata/telegraf/internal/limiter"
+	commonaliyun "github.com/influxdata/telegraf/plugins/common/aliyun"
```
Comment (srebhan) [inline plugins/inputs/aliyuncms/aliyuncms.go]:
```suggestion
	common_aliyun "github.com/influxdata/telegraf/plugins/common/aliyun"
```

File: plugins/inputs/aliyuncms/aliyuncms.go [hunk 2]
```diff
@@ -253,86 +216,135 @@ func (s *AliyunCMS) updateWindow(relativeTo time.Time) {
...
+	}
+	for k, v := range disc {
+		tags[k] = v
+	}
+	return true
+}
+
 // Gather given metric and emit error
-func (s *AliyunCMS) gatherMetric(acc telegraf.Accumulator, metricName string, metric *metric) error {
+func (s *AliyunCMS) gatherMetric(acc telegraf.Accumulator, metricName string, m *metric) error {
```
Comment (srebhan) [inline plugins/inputs/aliyuncms/aliyuncms.go:262]: Can we please revert the changes to `gatherMetric` to keep the diff small! If you want to refactor, please do this in a preparatory PR.

File: plugins/inputs/aliyuncms/README.md [hunk 3]
```diff
@@ -187,6 +188,11 @@ case](https://en.wikipedia.org/wiki/Snake_case)
   - {metric}_maximum     (metric Maximum value)
   - {metric}_value       (metric Value value)
 
+## RDS Performance metrics
+
+You can use the dedicated [aliyunrds](../aliyunrds/README.md) plugin to
+collect RDS Performance metrics.
+
```
Comment (srebhan) [inline plugins/inputs/aliyuncms/README.md:195]: I don't think you need to link that other plugin here...

File: plugins/inputs/aliyuncms/sample.conf [hunk 4]
```diff
@@ -39,6 +39,7 @@
   ## - acs_rds_dashboard
   ## - acs_slb_dashboard
   ## - acs_vpc_eip
+  ## Note: RDS Performance metrics are now supported via the dedicated aliyunrds plugin
```
Comment (srebhan) [inline plugins/inputs/aliyuncms/sample.conf:42]: Please remove!

**Comment by ZPascal** [comment]
> Thanks for the PR @ZPascal! This looks much better, however, I prefer the following steps each in a separate PR:
> 
> 1. Move the common part of the `aliyuncms` plugin into `common/aliyun` without adding the new plugin.
> 2. Migrate the common part and `aliyuncms` plugin to the [new Aliyun SDK](https://github.com/aliyun/alibabacloud-oss-go-sdk-v2) as the other one is EOL since March 2025.
> 3. Refactor the `aliyuncms` plugin if you wish to.
> 4. Add the new `aliyunrds` plugin.
> 
> What do you think? You can leave this PR here if you want and rebase it once the other steps are merged into master to gradually reduce the PR diff.

Sure, I can split the PR's.

File: plugins/inputs/aliyuncms/aliyuncms.go (hunk 1 above)
Comment (ZPascal) [reply plugins/inputs/aliyuncms/aliyuncms.go]: Sure

File: plugins/inputs/aliyuncms/aliyuncms.go (hunk 2 above)
Comment (ZPascal) [reply plugins/inputs/aliyuncms/aliyuncms.go:262]: I'll handle it in a refactoring PR.

File: plugins/inputs/aliyuncms/README.md (hunk 3 above)
Comment (ZPascal) [reply plugins/inputs/aliyuncms/README.md:195]: I think it is better to document it and make the end users aware of the new plugin.

File: plugins/inputs/aliyuncms/sample.conf (hunk 4 above)
Comment (ZPascal) [reply plugins/inputs/aliyuncms/sample.conf:42]: Okay, I'll handle it.

File: plugins/inputs/aliyuncms/README.md (hunk 3 above)
Comment (srebhan) [reply plugins/inputs/aliyuncms/README.md:195]: I don't like this because there is no clear definition on when to link plugins and when not. The user will be able to discover the new plugin either by the name or when browsing the [plugin list in the docs](https://docs.influxdata.com/telegraf/v1/plugins/)...

**Comment by ZPascal** [comment]
> 2. Migrate the common part and `aliyuncms` plugin to the [new Aliyun SDK](https://github.com/aliyun/alibabacloud-oss-go-sdk-v2) as the other one is EOL since March 2025.

Hi @srebhan, I have reviewed the potential migration between the SDKs and would like to address this in a separate PR. This is not a simple migration, but rather a completely new implementation.

**Comment by srebhan** [comment]
Sure, feel free to postpone the migration to another PR!

**Comment by srebhan** [comment]
Closing this due to inactivity in PR #18280.
